Java code parsing + Symbol resolving

nikolagudelj opened the pull request:
2 phase parsing.
First phase for class/interface/member/field lookup and storage.
Second phase for resolving references/member calls inside of methods and constructors.

Closes #1, Closes #3, Closes #5, Closes #7

Commit:
Feature: Outer visitor. ClassMap, Context, MemberSignature.

Commit:
Refactor: Parse multiple files. Name changes. Code refactoring.

Commit:
Feature: Demo resolver

For static and local methods.

Commit:
Feature: Local variable, param, field invocations.

Also removed 'internal' keywords from all files.

Commit:
Small refactor + method argument feature.

Commit:
Refactor: Code cleanup.

Commit:
Feature: Casting support. Object functions as params.

Commit:
Feature: Class hierarchy. Refactor.

Calling methods via super. Calling methods from parent. Method chaining works. New test .java classes added.

Commit:
Refactor: Classmap simplified with ContextHolder

Commit:
Refactor: Code cleanup.

Commit:
Feature: Notify usafe method in CadetNodeSolver.

Every registered field access or method call is now registered, instead of only top level ones. Code cleanup.

Commit:
Feature: Field access. WithCallerNode added.

Commit:
Diagnostics console print. Small fixes.

Commit:
Refactor: SymbolMap interface removed.

Commit:
Feature: HierarchyVisitor. New test package.

Commit:
Feature: Basic interface support.

Commit:
Feature: DI, Lisk sub. pr. resolver. #7 #5

Commit:
Refactor: Code cleanup and simplification.

Commit:
Fix: Constructor resolving failure.

Commit:
Feature: Null argument support.

Commit:
Refactor: Removal of SymbolContextMap pt1.

Commit:
Refactor: Removal of SymbolContextMap pt2

SymbolContextMap renamed to SymbolSolvingBundle. It is now passed to every SolverNode object.

Commit:
Refactor: Repackaging into FirstPass and SecondPass.

Commit:
Refactor: Renaming classes.

Commit:
Refactor: MemberSignature works with Hierarchy injection instead of static.

Commit:
Refactor: JavaCodeParser simplified.

Commit:
Refactor + Fix: Deleted SymbolContextMap

Fix: Wildcard return type is now assigned to every CadetSolverNode which doesn't have a return type defined after calling 'resolve()'. This prevents crashes when unknown argument types are given as arguments, and uses the Signature algorithm to potentially resolve known member calls with unknown argument types.

Luburic commented on src/main/kotlin/JavaCodeParser.kt:
```diff
@@ -0,0 +1,51 @@
+    private val resolverVisitor = SymbolResolverVisitor()
+    private val resolverResourceList = mutableListOf<ResolverVisitorResource>()
+
+    fun parseFiles(sourceCodeDtoList: List<SourceCodeDto>): List<CadetClass>? {
```
First, I would rename the function as it doesn't really parse files. "parseCode" or "parseSourceCode" might be more appropriate.

Since this function is the main API of the JavaCodeParser, it should be at a high level of abstraction. Its body should be a series of function calls, where each function call is a step in the parsing process, described by the significant name.

Using the [split loop](https://www.refactoring.com/catalog/splitLoop.html) refactoring, I would segregate the _for_ in line 23, where the first builds a list of compilation units and the second builds and calls the resolver visitor. The loops can then be extracted into separate functions that have a significant name.

My general issue with this method is the lack of clarity about what is going on. Maybe we could try the following to resolve this:
1) Write a description in 3-5 sentences that best describes what this function does. Try to be clear and concise, while avoiding generic words. Imagine that you are writing documentation or speaker's notes for a presentation (you will most likely reuse this description for your bachelor thesis and presentation). Feel free to write it in Serbian.
2) Once you are done, see if you can map each 1-2 sentences to a segment of the code.
3) Try removing as many words as possible while maintaining the general meaning. See that you remove everything that can be inferred from the class and calling method name.
4) What you have left is the function name for the step of the source code parsing.

nikolagudelj replied:
JavaCodeParser has been refactored. Each phase is contained within 1 single private method. Those methods are then called from the only public API method.

Luburic commented on src/main/kotlin/JavaCodeParser.kt:
```diff
@@ -0,0 +1,51 @@
+            catch (e: ParseProblemException) {
+                return null
+            }
+            resolverResourceList.add(
```
resolver, resource, and list are all pretty generic terms.

First, we don't really need "List" in the name, as this is inferred from the type. It is sufficient to call the variable resolverResource**s**. Next, we want to see what is meant by _Resource_ and to see if there is a better, less generic term. The same goes for _resolver_ - what would be a good word for this set of behavior, in Serbian or English? We can brainstorm this, but see if you can come up with some candidate solutions.

nikolagudelj replied:
ResolverResource is now removed. SymbolResolverVisitor works with 2 parameters, CompilationUnit and JavaPrototype.

Luburic commented on src/main/kotlin/cadet_model/CadetField.kt:
```diff
@@ -0,0 +1,14 @@
+    lateinit var parent: CadetClass
+    val modifiers = mutableListOf<CadetModifier>()
+
+    fun isStatic(): Boolean {
```
What behavior does this function support?

nikolagudelj replied:
Removed.

Luburic commented on src/main/kotlin/cadet_model/abs/CadetVariable.kt:
```diff
@@ -0,0 +1,6 @@
+package cadet_model.abs
+
+// Make this an abstract class?
+abstract class CadetVariable {
```
What would the hierarchy look like? Anything that has a type (e.g., function, class, variable, field) would inherit this?

If we can't think of some behavior to go along with the type, then we would replace a "string type" with a "CadetVariable type", which only increases any access chains (e.g., if method.type.type instead of method.type).

Luburic commented on test/tests/BaseTest.java:
```diff
@@ -0,0 +1,31 @@
+package test.tests;
+
+class BaseTest {
```
Be sure to run your parser against real open source projects. Use https://github.com/search/advanced to find popular (e.g., over 1000 or 5000 starts) open sources projects written in Java.

My experience is that you only really know if your parsers are working when you have tests with real code. After going through 10 such repositories I still occasionally find bugs due to different coding styles and code use.

Luburic commented on src/main/kotlin/first_pass/node_parser/MemberDeclarationParser.kt:
```diff
@@ -0,0 +1,46 @@
+        }
+    }
+
+    fun instantiateConstructor(node: ConstructorDeclaration, parent: CadetClass): CadetMember {
```
Do we have Accessor support? We can define these methods as those that start with is/get/set and access a field that shares part of the method's name (i.e., methodName.contains(fieldName, ignore_case)). While this algorithm is imperfect, it should detect most accessors.

Luburic commented on src/main/kotlin/hierarchy/HierarchyGraph.kt:
```diff
@@ -0,0 +1,150 @@
+        node.parent?.let { recursiveParentAdd(it, list) }
+    }
+
+    private fun instantiateObjectBase(): CadetClass {
```
What is the purpose of this function?

nikolagudelj replied:
This is development purposes only and will probably be removed later down the path.

Luburic commented on src/main/kotlin/second_pass/context/ClassContext.kt:
```diff
@@ -0,0 +1,11 @@
+
+open class ClassContext(val cadetClass: CadetClass) {
+
+    fun getField(name: String): CadetField? {
```
If I recall correctly, we talked about renaming the _Context_ to _Scope_. Will we go through with this?

Luburic commented on src/main/kotlin/second_pass/resolver/solver_nodes/abs/CadetSolverNode.kt:
```diff
@@ -0,0 +1,54 @@
+    private fun notifyContextOfUsage() {
+        if (resolvedReference == null) {
+            try {
+                returnType
```
I can't help but think of the blood mage's quote from warcraft "I smell magic in the air." :)

Put simply, the logic here is very unintuitive and does not express the intent. What is the purpose of this method? What does it do? Can we refactor to make it more clear?

Luburic commented on src/main/kotlin/second_pass/resolver/solver_nodes/abs/MemberCallSolverNode.kt:
```diff
@@ -0,0 +1,42 @@
+    SignableMember {
+    private val children = mutableListOf<BaseSolverNode>()
+
+    override fun doResolve() {
```
doResolve can be simplified to "resolve". We should discuss if this is the best name (and if it is, we should clearly define its meaning in our ubiquitous language).

Luburic commented on src/main/kotlin/second_pass/signature/SignableMember.kt:
```diff
@@ -0,0 +1,7 @@
+package second_pass.signature
+
+interface SignableMember {
+    fun getName(): String
```
I can't wrap my head around this set of classes related to the signature.

Do we really need so many, or is it enough to offer something along the lines of https://github.com/Clean-CaDET/platform/blob/master/RepositoryCompiler/CodeModel/CaDETModel/CodeItems/CaDETMember.cs#L20-L37

Luburic reviewed (commented):
First of all, I think the new packaging structure greatly improves the code organization. I've left several comments where I felt that there is room for improvement or where I had a question or two.

I still want to go over the code in more detail when I get the chance (by the end of the week) to see if I can identify more room for improvement and areas that are unclear. We can organize a sync after that.

Luburic commented on src/main/kotlin/JavaCodeParser.kt:
```diff
@@ -0,0 +1,51 @@
+    private val resolverVisitor = SymbolResolverVisitor()
+    private val resolverResourceList = mutableListOf<ResolverVisitorResource>()
+
+    fun parseFiles(sourceCodeDtoList: List<SourceCodeDto>): List<CadetClass>? {
```
I would also rename the parameter - it is clear that it is a DtoList from the type. Another question is - do we need the DTO or is it enough to send the text of the code? As far as I can tell, we are not using the fileName.

nikolagudelj replied:
Dto was imagined as development-only. Further down the path I intend to replace it with source code strings.

nikolagudelj replied:
SourceCodeDto has now been removed. JavaCodeParser takes in 1 parameter in the form of List<String>.

Luburic commented on src/main/kotlin/JavaCodeParser.kt:
```diff
@@ -0,0 +1,51 @@
+            )
+        }
+        resolverVisitor.resolveResources(hierarchyGraph, resolverResourceList)
+        return hierarchyGraph.getClasses()
```
Regarding the transformation of this function into high-level steps - from what I can tell, the steps are very closely related to the CSharpCodeParser https://github.com/Clean-CaDET/platform/blob/master/RepositoryCompiler/CodeModel/CodeParsers/CSharp/CSharpCodeParser.cs#L28-L32

With split loop we can get the following:
1) LoadCompilationUnits
2) CreateCaDETPrototypes
3) ConnectCaDETGraph
4) CalculateMetrics

As a side note, split loop might seem like performance loss but in this case it should be very negligible as the individual operations are far more taxing (on medium-to-large C# projects (over 10k functions) it takes between 30 to 60 seconds to parse the model on my machine, while the performance gain of reducing a loop or two is counted in microseconds).

Luburic replied:
As a side note (maybe save this idea to your Trello board or whatever you use for notetaking), we should have a slide or two that visualize and demonstrate how the CaDET object model changes after each step in the processing for your bachelor defense.

Luburic commented on src/main/kotlin/first_pass/ClassPrototypeVisitorOutput.kt:
```diff
@@ -0,0 +1,8 @@
+
+import cadet_model.CadetClass
+
+class ClassPrototypeVisitorOutput {
```
Would a better name for this be ClassPrototype?

nikolagudelj replied:
Yes, it would.

Luburic commented on src/main/kotlin/first_pass/node_parser/ClassDeclarationParser.kt:
```diff
@@ -0,0 +1,23 @@
+            this.parent = parent
+        }
+    }
+
```
I'm not sure what to think about these singletons. On the one hand, I think it is good that you separated the visiting logic from the parsing logic. On the other hand, this singleton is used in the first_pass and in the second_pass (the second method), which signals that are packages are not as cohesive as they should be.

The static classes in the node_parser could be acceptable if they were package-private and hidden from the rest of the app.

An alternative would be to have the DeclarationParser classes be instantiated in the visitor, but I'm not too keen on that solution either... What I did was have essentially two DeclarationParsers - one for the class (which included the fields) and one for the methods, which included your Member, MethodCall, and FieldAccess. I'm not too keen on that solution either 🙂.

I don't think we need a major refactoring here, but it would be good if we could disconnect any coupling between these static classes and the second_pass classes if it is not too big of a problem.

nikolagudelj replied:
This was my also my concern. The ClassDeclarationParser isn't the only one that's used in the 2nd pass. If it were only that, I could easily play around it. But the problem is, 2nd pass also works with nodes because it's a visitor, and with the current algorithm/architecture, during the Resolver visit, whenever the app bumps into a MethodCallExpression, FieldAccess, or some other node which needs resolving, I have to access those Node-parsing mechanisms in order to figure out what the current node represents (name, arguments, etc). So the problem goes much deeper and this current solution is definitely up for refactoring, but I left it this way simply because it got the job done.

nikolagudelj replied:
Singletons have been repackaged, and are only being used by visitors from their own package.

Luburic commented on src/main/kotlin/first_pass/node_parser/MemberDeclarationParser.kt:
```diff
@@ -0,0 +1,46 @@
+object MemberDeclarationParser : AbstractNodeParser() {
+
+    fun instantiateMethod(node: MethodDeclaration, parent: CadetClass): CadetMember {
+        return CadetMember().apply {
```
As far as I can tell we have duplicate code between the method and constructor. A function that takes as input CaDETMemberType would be nice.

nikolagudelj replied:
I did this on purpose because it avoided messy casting and if-statements. 
I extracted the shared parts into a separate private 'instantiateBaseMember' function, and left parts which are Node-specific to be handled by their respective functions.

Luburic commented on src/main/kotlin/first_pass/node_parser/FieldAccessExpressionParser.kt:
```diff
@@ -0,0 +1,27 @@
+package first_pass.node_parser
```
Should this class and the MethodCallExpressionParser be in the second_pass package?

As a side note (in case I didn't already leave a comment regarding this somewhere) we can rename first_pass to prototype_construction and second_pass to graph_connection or something along those lines.

nikolagudelj replied:
Yes, I haven't thought about this. I extracted all 2nd-pass-specific parsers into the second_pass package.

nikolagudelj replied:
All parsers are now repackaged.

Luburic commented on src/main/kotlin/JavaCodeParser.kt:
```diff
@@ -0,0 +1,51 @@
+    private fun createCadetPrototype(cUnit: CompilationUnit): CadetClass {
+        prototypeVisitor.parseTree(cUnit)
+            .let { output ->
+                hierarchyGraph.addClass(output.cadetClass)
```
How does this function change if we use split loot?

From what I can tell, we will have a List of ClassPrototypes as the result of the CreatePrototypes step.

We can feed this into the ConnectGraph step, which should encapsulate both the hierarchyGraph logic and the resolverVisitor logic. Looking at the code, I wonder if we can merge these two classes in some way?

Since the resolverVisitor.resolveResource already takes as input the hierarchyGraph, we can move the hierarcyhGraph.getClasses expression as the return value of the resolveResources function.

Next, we can have the resolveResource function accept the list of ClassPrototypes and from this list and the compilation units list it can create the hierarchyGraph and resolverResourceList - the question then becomes do we even need these classes as such or can this transformation along with additional refactoring bring a more elegant design?

nikolagudelj replied:
HierarchyVisitor has been removed. Its functionalities have been moved to ResolverVisitor. Also, JavaCodeParser at this moment, knows nothing about the existence of HierarchyGraph, it is a part of implementation details of the second pass stage.

Luburic commented on src/main/kotlin/second_pass/SymbolResolverVisitor.kt:
```diff
@@ -0,0 +1,103 @@
+
+    private fun initializeDataStructures(hierarchyGraph: HierarchyGraph) {
+        visitorContext = VisitorContext()
+        hierarchyVisitor = HierarchyVisitor(visitorContext, hierarchyGraph)
```
Is this a visitor? Visitors are known for their visit method. If this is not a visitor, than neither should the VisitorContext have the word Visitor in its name.

**KEY INSIGHT**
While on the subject, the VisitorContext class in this current design is pretty dangerous and bug-prone, as we rely on other classes (hierarchyVisitor, resolver) to change the state of this object as it is being passed around (usually as a side-effect of the function).

Since the hierarchyVisitor has a single method that is primarily a call to the static class, do we need it? Can we make the SymbolResolverVisitor focus on the visiting, while the SymbolResolver (which we might call CaDETGraphConnector?) focuses on the linking of the hierarchy, accessed fields, and invoked methods?

If we follow the same pattern for the first_pass, we will have some node_connector static classes that will actually be the FieldAccessParser, MethodCallParser, and HierarchyParser (that contains the function from the current ClassDeclaration parser).

Luburic commented on src/main/kotlin/second_pass/SymbolResolverVisitor.kt:
```diff
@@ -0,0 +1,103 @@
+    }
+
+// Resolving methods
+    fun resolveResources(hierarchyGraph: HierarchyGraph, resources: List<ResolverVisitorResource>) {
```
Regarding _ResolverVisitorResource_ - what would happen to the design if this pairing of the compilation unit and associated cadetclass did not exist? How would it change?

nikolagudelj replied:
ResolverVisitorResource does not exist anymore, it has been replaced with 2 parameters (which were contained within this dto anyways).

Luburic reviewed (changes requested):
I did another passthrough and got a much better understanding of the flow. I've left comments that should help with redesigning the most complicated parts of the code.

We will need to discuss the solver package in more detail once we go over everything that leads up to it.

Commit:
Refactor: HierarchyGraph, and some smaller changes all over

All recursive functions removed from hierarchy graph, replaced with while loops.

Commit:
Refactor: HierarchyGraph. JavaCodeParser. JavaPrototype

HierarchyGraph removed. JavaCodeParser methods broken up into smaller pieces where each method represents 1 phase in parsing/resolving. Introduced JavaPrototype to handle Interfaces and Classes through 1 mutual interface. Removed SourceCodeDto and ResolverResource.

Commit:
Refactor: HierarchyGraphFactory

Commit:
Feature: Inner class support. Refactor.

Commit:
Feature: EnclosedExpr support

Commit:
Refactor: Transformed ANP to a static utility AstNodeUtil

Commit:
Parsing refactor + small enum bug fixes

Commit:
Gitignore update

Commit:
Refactor: Signature and VisitorContext (removed)

Commit:
Refactor: ResolverWizard

Removed query functionalities from SymbolResolver, and moved them over to a new class, ResolverWizard.
